### PR TITLE
[2201.2.x] Fix "pattern will not be matched" in match statement when iterating over characters of string in foreach loop."

### LINKED ISSUE #36710: The compiler warning "pattern will not be matched" in `match` statement when iterating over characters of string in `foreach` loop.
**Description:**
The compiler warning "pattern will not be matched" in `match` statement when iterating over characters of string in `foreach` loop. However the program seems to work just fine.

**Steps to reproduce:**

```
$ cat issue.bal 
import ballerina/io;
public function main() {
    string input = "1x2";

    foreach string item in input {
        match item {
            "1" => { io:println("ONE"); }
            "2" => { io:println("TWO"); }
            _ => { io:println("SOMETHING ELSE"); }
        }
    }
}
```

```
$ bal run issue.bal 
Compiling source
	issue.bal
WARNING [issue.bal:(7:13,7:16)] pattern will not be matched
WARNING [issue.bal:(8:13,8:16)] pattern will not be matched

Running executable

ONE
SOMETHING ELSE
TWO
```

**Affected Versions:**

```
$ bal version
Ballerina 2201.1.0 (Swan Lake Update 1)
Language specification 2022R2
Update Tool 1.3.9
```

### PR DESCRIPTION
**Pull request by gimantha**
Fixes #36710

### CODE REVIEW

File: tests/jballerina-unit-test/src/test/resources/test-src/statements/matchstmt/const-pattern.bal [hunk 1]
```diff
@@ -688,6 +688,68 @@ function constPatternWithPredeclaredPrefix2(int[] x) returns int {
...
+    }
+    assertEquals("OtherTwoTwo", output);
+}
+
+function testMatchStmtInsideForeach() {
+    testMatchStmtInsideForeachString1();
+    testMatchStmtInsideForeachString2();
+    testMatchStmtInsideForeachInt();
+}
+
```
Comment (SasinduDilshara) [inline tests/jballerina-unit-test/src/test/resources/test-src/statements/matchstmt/const-pattern.bal:752]: Shall we add one test with constant variables.
ex:- 
```
function testMatchStmtInsideForeachString1() {
    string input = "1x2";
    string output = "";
    const string one = "1";
    const string two = "2";
    
    foreach string char in input {
        match char {
            one => {
                output = output.concat("One");
            }
            two => {
                output = output.concat("Two");
            }
            _ => {
                output = output.concat("Other");
            }
        }
    }
    assertEquals("OneOtherTwo", output);
}
```

File: tests/jballerina-unit-test/src/test/resources/test-src/statements/matchstmt/const-pattern.bal [hunk 2]
```diff
@@ -688,6 +688,68 @@ function constPatternWithPredeclaredPrefix2(int[] x) returns int {
...
+    assertEquals("OneOtherTwo", output);
+}
+
+function testMatchStmtInsideForeachInt() {
+    int:Signed8[] input = [7, 2, 2];
+    string output = "";
+    foreach int number in input {
+        match number {
+            2 => {
+                output = output.concat("Two");
```
Comment (SasinduDilshara) [inline tests/jballerina-unit-test/src/test/resources/test-src/statements/matchstmt/const-pattern.bal:737]: Isn't it better if we use `output += "Two";`.
Using `concat` means we depend on `langlib string` module for this test.

Comment (gimantha) [reply tests/jballerina-unit-test/src/test/resources/test-src/statements/matchstmt/const-pattern.bal:737]: `langlib string` is a builtin module we can use. we can depend on that. I don't see a any specific reason not to

File: tests/jballerina-unit-test/src/test/resources/test-src/statements/matchstmt/const-pattern.bal (hunk 1 above)
Comment (gimantha) [reply tests/jballerina-unit-test/src/test/resources/test-src/statements/matchstmt/const-pattern.bal:752]: will add

Comment (gimantha) [reply tests/jballerina-unit-test/src/test/resources/test-src/statements/matchstmt/const-pattern.bal:776]: Added in https://github.com/ballerina-platform/ballerina-lang/pull/37334/commits/a520b494cfd7b6ba1c35380a6dfa2ee7ca03ba01

File: tests/ballerina-compiler-api-test/src/test/resources/test-src/statements/symbols_in_reg_compound_stmt.bal [hunk 3]
```diff
@@ -45,7 +45,7 @@ function testForeach() {
         call();
     }
 
-    foreach int[] [a, b, c] in [[1, 2, 3], [3, 4, 5]] {
+    foreach int[3] [a, b, c] in [[1, 2, 3], [3, 4, 5]] {
```
Comment (MaryamZi) [inline tests/ballerina-compiler-api-test/src/test/resources/test-src/statements/symbols_in_reg_compound_stmt.bal:48]: Didn't we decide to clarify this in the spec?

File: tests/jballerina-unit-test/src/test/resources/test-src/statements/matchstmt/const-pattern.bal [hunk 4]
```diff
@@ -688,6 +688,92 @@ function constPatternWithPredeclaredPrefix2(int[] x) returns int {
...
+    }
+    assertEquals("OneOtherTwo", output);
+}
+
+function testMatchStmtInsideForeach() {
+    testMatchStmtInsideForeachString1();
+    testMatchStmtInsideForeachString2();
+    testMatchStmtInsideForeachInt();
+    testMatchStmtInsideForeachWithConst();
+}
```
Comment (KavinduZoysa) [inline tests/jballerina-unit-test/src/test/resources/test-src/statements/matchstmt/const-pattern.bal:775]: Shall we add this type of test too,
```
    string s = "ABCDEF";
    foreach string:Char item in s {
        
    }
```

**Review by KavinduZoysa — APPROVED** [review]

**Pull request merged**